site/content/ytt/docs/develop/how-to-modularize.md

```diff
@@ -0,0 +1,481 @@
+#@ load("format.star", "name")
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: #@ name(name, "deployment")
+  namespace: #@ data.values.namespace
+  labels: #@ labels(data.values.name, data.values.version)
+spec:
+  replicas: #@ data.values.replicas
+  template:
+    spec:
+      containers:
+        - name:  #@ name(name, "service")
+          image: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: #@ name(data.values.name, "service")
+  labels: #@ labels(data.values.name, version)
+spec:
+  type: ClusterIP
+  ports:
+    - port: 38080
+      targetPort: 8080
+```
+
+```yaml
+#@data/values-schema
+---
+name: "frontend"
+namespace: "default"
+replicas: 1
+version: 0.1.0
+```
+```shell
+$ tree .
+.
+├── config.yml
+├── schema.yml
+├── format.star
+└── labels.lib.yml
+```
+
+Execute ytt via `ytt -f config.yml -f schema.yml -f labels.lib.yml -f format.star`
+
+## Extract functionality into custom library
+You can extract a whole set of input files (i.e. templates, overlays, data values, etc.) into a "Library" in the `_ytt_lib/` folder. A library can be thought of as a separate self-contained ytt invocation. Libraries are _not_ automatically included in `ytt` output. They must be programmatically imported using the [`library` module](/ytt/docs/develop/lang-ref-ytt-library/), configured, evaluated, and inserted into a template that is part of the output.
+
+### Uses of a custom library
+Libraries are helpful when
+* importing 3rd party configuration into one combined piece of configuration. Such as from a division of responsibility or shared configuration.
+  * For example, having a library that provides helpful functions that needs to be used across multiple teams. Authors may use a tool to ensure these stay in sync.
+* A template is needed by two distinct applications like frontend and backend.
+* There is a need to update one application with an evaluated value from the other. [Playground example here](/ytt/#example:example-ytt-library-module).
+
+All the previous section's files have moved to `_ytt_lib/resources`:
+
+```shell
+config/
+$ tree .
+├── _ytt_lib/
+│   └── resources/
+│       ├── config.yml
+│       ├── schema.yml
+│       ├── labels.lib.yml
+│       └── format.star
+├── config.yml
+└── values.yml
+```
+In this example we want the resources from the library for a frontend application, and also for a backend application. We will use this library to create both of these.
+
+Focusing on only the two top level files, `config.yml`, and `values.yml`, we import the custom library, provide it data values for a frontend and separately for a backend, and use `#@ template.replace()` to insert it inline so it shows in the output.
+
+```yaml
+#! config.yml
+#@ load("@ytt:data", "data")
+#@ load("@ytt:library", "library")
+#@ load("@ytt:template", "template")
+
+#@ resources_lib = library.get("resources")
+
```

Comment:
nit: i would argue to kill this newline and next newline. to make `resources_lib` + `backend` + `frontend` seem to be a group.